### PR TITLE
fix: Reset image error flag when imageUrl or defaultImageUrl is changed in Image widget

### LINKED ISSUE #11494: [Bug]: Image URL stops working after I remove default image URL from image widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

[LOOM DEMO](https://www.loom.com/share/e031a1bb34ed4674a8958372b88e1595)

### Steps To Reproduce

1. Drag and drop image widget
2. Paste an image URL
3. remove the default Image url
4. Now if I copy paste the existing image URL it stops working

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11768: [Bug]: URLs ( containing binding value  ) doesn't work in the "Image field", rather works in the "Default Image field"
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When I try to use an Image URL that contain binding values for ex `https://images.unsplash.com/photo-{{table.selectedRow.value}}e72b0397e57bixlib=rb1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1335&q=80` in the Image field section it doesnt work. When I empty that field and use it in the Default image then it works.

### Steps To Reproduce

1. Go to "Image field" section of Image widget
2. Enter the URL with binding value in it `https://images.unsplash.com/photo-{{table.selectedRow.value}}-e72b0397e57b?ixlib=rb1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1335&q=80g`
3. Image doesnt show up
4. Now remove that URL from that section , empty it
5. Paste that URL in the "Default Image" section
6. It works in this way

### Public Sample App

https://app.appsmith.com/applications/62051a456b4b1e154a398053/pages/62051a456b4b1e154a398056

### Version

Cloud - Appsmith v1.6.14-SNAPSHOT

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Image URL stops working after removing default image URL from image widget to fix this issue here we set`imageError` set to false when DefaultUrl or imageUrl is set.

Fixes #11494 
Fixes #11768

## Type of change

- Bugfix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/handle-imageerror-flag 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.87 **(-0.01)** | 37.23 **(-0.01)** | 35.22 **(0)** | 56.18 **(-0.01)**
 :red_circle: | app/client/src/selectors/commentsSelectors.ts | 83.61 **(-1.64)** | 61.76 **(-2.95)** | 73.33 **(0)** | 88.24 **(-2.35)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 88.98 **(-0.78)** | 70.59 **(0)** | 100 **(0)** | 92.38 **(-0.95)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :red_circle: | app/client/src/widgets/ImageWidget/component/index.tsx | 18.75 **(-0.6)** | 13.95 **(-0.86)** | 9.09 **(-0.43)** | 20.99 **(-0.8)**</details>

### CODE REVIEW

File: app/client/src/widgets/ImageWidget/component/index.tsx [hunk 1]
```diff
@@ -116,6 +116,18 @@ class ImageComponent extends React.Component<
     };
   }
 
+  componentDidUpdate = (prevProps: ImageComponentProps) => {
+    // imageError flag set true but not getting reset if imageUrl and defaultImageUrl are empty.
```
Comment (sbalaji1192) [inline app/client/src/widgets/ImageWidget/component/index.tsx:120]: Use multiLine syntax

Comment (sbalaji1192) [reply app/client/src/widgets/ImageWidget/component/index.tsx:120]: `reset the imageError flag when the defaultImageUrl or imageUrl changes` should suffice.

**Review by sbalaji1192 — COMMENTED** [review]
Let's add a cypress case to test this scenario. Comment out your fix and make sure that cypress is catching the issue.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  it("Add new image and check image src", function() {
+    cy.testCodeMirror(this.data.NewImage);
+    // if imageError flag not reset properly, this test will fail.
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.NewImage);
+  });
+});
+afterEach(() => {
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:55]: remove

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    cy.addDsl(dsl);
+  });
+
+  it("Check default image and new image", function() {
+    cy.openPropertyPane("imagewidget");
+    cy.testJsontext("defaultimage", this.data.defaultimage);
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.defaultimage);
+    cy.wait(1000);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:18]: Do we need to wait for 1 sec ?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @param{URL} ImageUrl
+     */
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:21]: out of place comment

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      `${viewWidgetsPage.imageWidget} div[data-testid=styledImage]`,
+    ).should("exist");
+
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", "");
+  });
+
+  it("Add new image and check image src", function() {
+    cy.testCodeMirror(this.data.NewImage);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:48]: Could you confirm that this case fails without your fix ?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      .invoke("attr", "src")
+      .should("contain", this.data.defaultimage);
+    cy.wait(1000);
+    /**
+     * @param{URL} ImageUrl
+     */
+    cy.testCodeMirror(this.data.NewImage);
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.NewImage);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:25]: Could you check that imageUrl supersedes the defaultImageUrl ?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      .should("contain", "");
+  });
+
+  it("Add new image and check image src", function() {
+    cy.testCodeMirror(this.data.NewImage);
+    // if imageError flag not reset properly, this test will fail.
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.NewImage);
+  });
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:53]: Also assert that error container doesn't exist.

**Comment by somangshu** [comment]
@techbhavin please have a detailed PR desc, Any body seeing it will not understand what have changed and why it changed making it to difficult to prepare release notes, please make sure that the PR desc quality persist even in the future.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js (hunk 5 above)
Comment (techbhavin) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:46]: Yes , Here test case is failing without new changes

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.defaultimage);
+
+    cy.testCodeMirror(this.data.NewImage);
+    cy.get(viewWidgetsPage.imageinner)
+      .invoke("attr", "src")
+      .should("contain", this.data.NewImage)
+      .should("not.contain", this.data.defaultimage);
+    cy.closePropertyPane();
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Image_validation_spec.js:24]: 1. Assert that when you add/change  image, viewWidgetsPage.imageinner src is pointing to the correct value
2. Assert that when you add defaultImage and image, viewWidgetsPage.imageinner src has the image URL. not defaultImage Url.

**Review by sbalaji1192 — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin Tested this PR and it is working as expected

**Comment by vivekverma2312** [comment]
@techbhavin Tested and it's working fine.

**Pull request merged**